File: day04/aoc4-2.py
#!/usr/bin/env python3

INPUT_FILE = "aocdata.txt"

DIRS = [(-1,-1), (0,-1), (1,-1),
        (-1, 0),          (1, 0),
        (-1, 1), (0, 1),  (1, 1)]

def read_grid(path):
    with open(path, 'r', encoding='utf-8') as f:
        lines = [line.rstrip('\n') for line in f if line.strip() != '']
    return [list(line) for line in lines]

def count_neighbors(grid, x, y):
    h, w = len(grid), len(grid[0])
    cnt = 0
    for dx, dy in DIRS:
        nx, ny = x + dx, y + dy
        if 0 <= nx < w and 0 <= ny < h and grid[ny][nx] == '@':
            cnt += 1
    return cnt

def remove_accessible(grid):
    h, w = len(grid), len(grid[0])
    to_remove = []
    for y in range(h):
        for x in range(w):
            if grid[y][x] == '@':
                if count_neighbors(grid, x, y) < 4:
                    to_remove.append((x, y))
    for x, y in to_remove:
        grid[y][x] = '.'
    return len(to_remove)

def main():
    grid = read_grid(INPUT_FILE)
    total_removed = 0
    while True:
        removed = remove_accessible(grid)
        if removed == 0:
            break
        total_removed += removed
    print(f"Total rolls removed: {total_removed}")

if __name__ == "__main__":
    main()

File: day04/aoc4-2.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $INPUT_FILE = "aocdata.txt";

my @DIRS = ([-1,-1],[0,-1],[1,-1],
            [-1, 0],        [1, 0],
            [-1, 1],[0, 1],[1, 1]);

sub read_grid {
    my ($path) = @_;
    open my $fh, '<', $path or die "Cannot open $path: $!";
    my @grid = ();
    while (my $line = <$fh>) {
        chomp $line;
        next if $line =~ /^\s*$/;
        push @grid, [split //, $line];
    }
    close $fh;
    return \@grid;
}

sub count_neighbors {
    my ($grid, $x, $y) = @_;
    my $h = scalar(@$grid);
    my $w = scalar(@{$grid->[0]});
    my $cnt = 0;
    for my $d (@DIRS) {
        my ($dx, $dy) = @$d;
        my $nx = $x + $dx;
        my $ny = $y + $dy;
        next if $nx < 0 || $ny < 0 || $nx >= $w || $ny >= $h;
        $cnt++ if $grid->[$ny][$nx] eq '@';
    }
    return $cnt;
}

sub remove_accessible {
    my ($grid) = @_;
    my $h = scalar(@$grid);
    my $w = scalar(@{$grid->[0]});
    my @to_remove;
    for my $y (0..$h-1) {
        for my $x (0..$w-1) {
            if ($grid->[$y][$x] eq '@') {
                if (count_neighbors($grid, $x, $y) < 4) {
                    push @to_remove, [$x, $y];
                }
            }
        }
    }
    for my $pos (@to_remove) {
        my ($x, $y) = @$pos;
        $grid->[$y][$x] = '.';
    }
    return scalar(@to_remove);
}

sub main {
    my $grid = read_grid($INPUT_FILE);
    my $total_removed = 0;
    while (1) {
        my $removed = remove_accessible($grid);
        last if $removed == 0;
        $total_removed += $removed;
    }
    print "Total rolls removed: $total_removed\n";
}

main();
